### PR TITLE
Auto uploading logs with mclo.gs link into dump

### PR DESCRIPTION
**Pull request by Jens-Co**
Leave it as draft since I need to figure out pathing for bungee/waterfall

### CODE REVIEW

File: connector/src/main/java/org/geysermc/connector/bootstrap/GeyserBootstrap.java [hunk 1]
```diff
@@ -120,4 +119,8 @@ default String getMinecraftServerVersion() {
     default SocketAddress getSocketAddress() {
         return null;
     }
+
+    default String logsPath() {
+        return "logs/latest.log";
```
Comment (Konicai) [inline connector/src/main/java/org/geysermc/connector/bootstrap/GeyserBootstrap.java:124]: I would make this return a Path instead of String

Also, since the default is "logs/latest.log", you don't need to override it on platforms that have their latest log in that location. You could mention in the javadoc that the default provides a Path of "logs/latest.log"

File: connector/src/main/java/org/geysermc/connector/command/defaults/DumpCommand.java [hunk 2]
```diff
@@ -74,7 +75,8 @@ public void execute(GeyserSession session, CommandSender sender, String[] args)
                     case "offline":
                         offlineDump = true;
                         break;
-
+                    case "logs":
+                        logsDump = true;
```
Comment (Konicai) [inline connector/src/main/java/org/geysermc/connector/command/defaults/DumpCommand.java:79]: Its common practice to still place a `break;` at the end

File: connector/src/main/java/org/geysermc/connector/command/defaults/DumpCommand.java [hunk 3]
```diff
@@ -141,6 +143,6 @@ public void execute(GeyserSession session, CommandSender sender, String[] args)
 
     @Override
     public List<String> getSubCommands() {
-        return Arrays.asList("offline", "full");
+        return Arrays.asList("offline", "full", "logs");
```
Comment (Konicai) [inline connector/src/main/java/org/geysermc/connector/command/defaults/DumpCommand.java:146]: imo it wouldn't hurt to also have "log" as a case, just as a catch for people who make a typo/forget

File: connector/src/main/java/org/geysermc/connector/dump/DumpInfo.java [hunk 4]
```diff
@@ -188,6 +195,25 @@ public DumpInfo() {
         }
     }
 
+    @Getter
+    public static class LogsInfo {
+        private String logs;
```
Comment (Konicai) [inline connector/src/main/java/org/geysermc/connector/dump/DumpInfo.java:200]: maybe `link`?

File: connector/src/main/java/org/geysermc/connector/dump/DumpInfo.java [hunk 5]
```diff
@@ -50,12 +53,15 @@
...
 import java.net.UnknownHostException;
+import java.nio.file.Path;
+import java.nio.file.Paths;
 import java.util.Properties;
 
 @Getter
 public class DumpInfo {
     @JsonIgnore
     private static final long MEGABYTE = 1024L * 1024L;
+    public static String LogsInfo;
```
Comment (Konicai) [inline connector/src/main/java/org/geysermc/connector/dump/DumpInfo.java:64]: this var doesn't seem used

File: connector/src/main/java/org/geysermc/connector/command/defaults/DumpCommand.java [hunk 6]
```diff
@@ -48,6 +48,7 @@
     private final GeyserConnector connector;
     private static final ObjectMapper MAPPER = new ObjectMapper();
     private static final String DUMP_URL = "https://dump.geysermc.org/";
+    public static boolean logsDump = false;
```
Comment (Konicai) [inline connector/src/main/java/org/geysermc/connector/command/defaults/DumpCommand.java:51]: Since this value depends on the instance of DumpCommand, it shouldn't be static (I would also make it private)

I would just add a createLogs boolean to DumpInfo() and LogsInfo()

**Review by Konicai — CHANGES_REQUESTED** [review]

File: bootstrap/standalone/src/main/java/org/geysermc/platform/standalone/GeyserStandaloneBootstrap.java [hunk 7]
```diff
@@ -284,8 +284,8 @@ public BootstrapDumpInfo getDumpInfo() {
     }
 
     @Override
-    public String[] logsPath() {
-        return new String[]{"logs/", "latest.log"};
```
Comment (lukeeey) [inline bootstrap/standalone/src/main/java/org/geysermc/platform/standalone/GeyserStandaloneBootstrap.java]: Shouldn’t this be getLogsPath to be consistent?

File: connector/src/main/java/org/geysermc/connector/dump/DumpInfo.java [hunk 8]
```diff
@@ -64,9 +68,13 @@
     private final Object2IntMap<DeviceOs> userPlatforms;
     private final HashInfo hashInfo;
     private final RamInfo ramInfo;
+    private LogsInfo logsInfo;
```
Comment (Konicai) [inline connector/src/main/java/org/geysermc/connector/dump/DumpInfo.java:71]: this should be final

File: connector/src/main/java/org/geysermc/connector/dump/DumpInfo.java [hunk 9]
```diff
@@ -64,9 +68,13 @@
...
     private final BootstrapDumpInfo bootstrapInfo;
 
-    public DumpInfo() {
+    public DumpInfo(boolean addLog) {
+        if (addLog) {
+            this.logsInfo = new LogsInfo(addLog);
+        }
```
Comment (Konicai) [inline connector/src/main/java/org/geysermc/connector/dump/DumpInfo.java:77]: could this be moved down between 
```java
this.ramInfo = new DumpInfo.RamInfo();

this.userPlatforms = new Object2IntOpenHashMap<>();
```
just so that order of DumpInfo() lines up with the order of the fields

Comment (Konicai) [reply connector/src/main/java/org/geysermc/connector/dump/DumpInfo.java]: Oh also, if you check addLog before creating LogsInfo, you don't need to pass addLog to LogsInfo

Comment (Jens-Co) [reply connector/src/main/java/org/geysermc/connector/dump/DumpInfo.java]: 😂

**Comment by Jens-Co** [comment]
I consider this pr done, if any more things need to be changed please let me know. I've tested it multiple times on various servers/forks.

File: bootstrap/bungeecord/src/main/java/org/geysermc/platform/bungeecord/GeyserBungeePlugin.java [hunk 10]
```diff
@@ -171,6 +172,16 @@ public BootstrapDumpInfo getDumpInfo() {
         return new GeyserBungeeDumpInfo(getProxy());
     }
 
+    @Override
+    public Path getLogsPath() {
+        Path getFork = Paths.get("proxy.log.0");
```
Comment (Camotoy) [inline bootstrap/bungeecord/src/main/java/org/geysermc/platform/bungeecord/GeyserBungeePlugin.java:177]: Check for BungeeCord fork status first, then get this path. I.E. `if (isBungeeCord) Paths.get(proxylog) else Paths.get(latest.log)`.

File: bootstrap/sponge/src/main/java/org/geysermc/platform/sponge/GeyserSpongePlugin.java [hunk 11]
```diff
@@ -168,4 +168,5 @@ public BootstrapDumpInfo getDumpInfo() {
     public String getMinecraftServerVersion() {
         return Sponge.getPlatform().getMinecraftVersion().getName();
     }
+
```
Comment (Camotoy) [inline bootstrap/sponge/src/main/java/org/geysermc/platform/sponge/GeyserSpongePlugin.java:171]: Revert this unnecessary diff.

File: connector/src/main/java/org/geysermc/connector/bootstrap/GeyserBootstrap.java [hunk 12]
```diff
@@ -120,4 +119,9 @@ default String getMinecraftServerVersion() {
     default SocketAddress getSocketAddress() {
         return null;
     }
+
+    default Path getLogsPath() {
+
```
Comment (Camotoy) [inline connector/src/main/java/org/geysermc/connector/bootstrap/GeyserBootstrap.java:124]: Unnecessary whitespace.

File: connector/src/main/java/org/geysermc/connector/bootstrap/GeyserBootstrap.java [hunk 13]
```diff
@@ -25,22 +25,21 @@
...
 import org.geysermc.connector.network.translators.world.WorldManager;
+import org.geysermc.connector.ping.IGeyserPingPassthrough;
 
 import javax.annotation.Nullable;
 import java.net.SocketAddress;
 import java.nio.file.Path;
+import java.nio.file.Paths;
 
 public interface GeyserBootstrap {
-
```
Comment (Camotoy) [inline connector/src/main/java/org/geysermc/connector/bootstrap/GeyserBootstrap.java:41]: Revert these changes, please.

File: connector/src/main/java/org/geysermc/connector/command/defaults/DumpCommand.java [hunk 14]
```diff
@@ -74,7 +75,10 @@ public void execute(GeyserSession session, CommandSender sender, String[] args)
                     case "offline":
                         offlineDump = true;
                         break;
-
+                    case "logs":
```
Comment (Camotoy) [inline connector/src/main/java/org/geysermc/connector/command/defaults/DumpCommand.java:78]: My opinion here is that we stick to one option here (remove the `log` case), as no one is going to know about that as an option unless we say it, and then we have two redundant switches floating around. Open to discussion as Konica brought this up.

File: connector/src/main/java/org/geysermc/connector/dump/DumpInfo.java [hunk 15]
```diff
@@ -99,12 +102,15 @@ public DumpInfo() {
...
+
         this.userPlatforms = new Object2IntOpenHashMap<>();
         for (GeyserSession session : GeyserConnector.getInstance().getPlayers()) {
             DeviceOs device = session.getClientData().getDeviceOs();
             userPlatforms.put(device, userPlatforms.getOrDefault(device, 0) + 1);
         }
-
```
Comment (Camotoy) [inline connector/src/main/java/org/geysermc/connector/dump/DumpInfo.java:107]: Revert this, please.

File: connector/src/main/java/org/geysermc/connector/dump/DumpInfo.java [hunk 16]
```diff
@@ -188,6 +194,25 @@ public DumpInfo() {
         }
     }
 
+    @Getter
+    public static class LogsInfo {
+        private String Link;
```
Comment (Camotoy) [inline connector/src/main/java/org/geysermc/connector/dump/DumpInfo.java:199]: This should be lowercase `link`.

File: connector/src/main/java/org/geysermc/connector/dump/DumpInfo.java [hunk 17]
```diff
@@ -188,6 +194,25 @@ public DumpInfo() {
...
+
+        public LogsInfo(Boolean addLog) {
+
+            try {
```
Comment (Camotoy) [inline connector/src/main/java/org/geysermc/connector/dump/DumpInfo.java:203]: Check for `addLog` first, then create the try/catch.

**Review by Camotoy — CHANGES_REQUESTED** [review]
A lot of this is nitpick stuff!!

The MCLogs API is a very lightweight dependency, so I'm fine with having it in Geyser. However, it does use Gson when a majority of Geyser processes use Jackson for JSON deserialization. I'm pretty sure Floodgate 2.0/Cumulus uses Gson, and as of right now there is no Jackson library for Adventure.

File: connector/src/main/java/org/geysermc/connector/command/defaults/DumpCommand.java (hunk 14 above)
Comment (Konicai) [reply connector/src/main/java/org/geysermc/connector/command/defaults/DumpCommand.java:78]: Honestly I'm indifferent, the majority of people are only going to know the subcommand if we ask them to run it

File: connector/src/main/java/org/geysermc/connector/dump/DumpInfo.java (hunk 17 above)
Comment (Konicai) [reply connector/src/main/java/org/geysermc/connector/dump/DumpInfo.java:203]: `addLog` shouldn't even be passed here, since its checked already in `DumpInfo()`. Or don't check in `DumpInfo()` and only check here.

Also, use `boolean` instead of `Boolean`

File: connector/src/main/java/org/geysermc/connector/dump/DumpInfo.java (hunk 8 above)
Comment (Konicai) [reply connector/src/main/java/org/geysermc/connector/dump/DumpInfo.java:69]: This should still be final :P

File: connector/src/main/java/org/geysermc/connector/utils/WebUtils.java [hunk 18]
```diff
@@ -132,4 +136,41 @@ private static String connectionToString(HttpURLConnection con) throws IOExcepti
...
+        http.setDoOutput(true);
+        //convert log to application/x-www-form-urlencoded
+        String content = "content=" + new BufferedReader(new InputStreamReader(Files.newInputStream(log.toRealPath()))).lines().collect(Collectors.joining("\n"));
+        byte[] out = content.getBytes(StandardCharsets.UTF_8);
+        int length = out.length;
+
+        //send log to api
+        http.setFixedLengthStreamingMode(length);
+        http.setRequestProperty("Content-Type", "application/x-www-form-urlencoded; charset=UTF-8");
+        http.setRequestProperty("User-Agent", userAgent + "/" + version + "/" + mcversion);
```
Comment (Konicai) [inline connector/src/main/java/org/geysermc/connector/utils/WebUtils.java:163]:
```suggestion
        http.setRequestProperty("User-Agent", "Geyser-" + GeyserConnector.getInstance().getPlatformType().toString() + "/" + GeyserConnector.VERSION);
```

File: connector/src/main/java/org/geysermc/connector/utils/WebUtils.java [hunk 19]
```diff
@@ -132,4 +136,41 @@ private static String connectionToString(HttpURLConnection con) throws IOExcepti
...
+        String userAgent = "unknown";
+        String version = "unknown";
+        //connect to api
+        URL url = new URL("https://api.mclo.gs/1/log");
+        URLConnection con = url.openConnection();
+        HttpURLConnection http = (HttpURLConnection) con;
+        http.setRequestMethod("POST");
+        http.setDoOutput(true);
+        //convert log to application/x-www-form-urlencoded
+        String content = "content=" + new BufferedReader(new InputStreamReader(Files.newInputStream(log.toRealPath()))).lines().collect(Collectors.joining("\n"));
```
Comment (Konicai) [inline connector/src/main/java/org/geysermc/connector/utils/WebUtils.java:156]:
```suggestion
        String content = "content=" + new BufferedReader(new InputStreamReader(Files.newInputStream(log.toRealPath()))).lines().collect(Collectors.joining());
```

Since its json, newline chars don't have to be added

File: connector/src/main/java/org/geysermc/connector/utils/WebUtils.java [hunk 20]
```diff
@@ -132,4 +136,41 @@ private static String connectionToString(HttpURLConnection con) throws IOExcepti
 
         return content.toString();
     }
+
+    /**
+     * Get the server log file and uploads it to mclo.gs
+     *
+     * @param log File to fetch
+     */
+    public static String mcLogs(Path log) throws IOException {
```
Comment (Camotoy) [inline connector/src/main/java/org/geysermc/connector/utils/WebUtils.java:145]: Change to `postLogs` or something?

File: connector/src/main/java/org/geysermc/connector/utils/WebUtils.java [hunk 21]
```diff
@@ -132,4 +136,41 @@ private static String connectionToString(HttpURLConnection con) throws IOExcepti
...
+        String mcversion = "unknown";
```
Comment (Camotoy) [inline connector/src/main/java/org/geysermc/connector/utils/WebUtils.java:146]: Placeholders? It doesn't appear you update these.

File: connector/src/main/java/org/geysermc/connector/utils/WebUtils.java [hunk 22]
```diff
@@ -132,4 +136,41 @@ private static String connectionToString(HttpURLConnection con) throws IOExcepti
...
+        String userAgent = "unknown";
+        String version = "unknown";
+        //connect to api
+        URL url = new URL("https://api.mclo.gs/1/log");
+        URLConnection con = url.openConnection();
+        HttpURLConnection http = (HttpURLConnection) con;
+        http.setRequestMethod("POST");
+        http.setDoOutput(true);
+        //convert log to application/x-www-form-urlencoded
```
Comment (Camotoy) [inline connector/src/main/java/org/geysermc/connector/utils/WebUtils.java:155]: Use `// Convert[...]` style formatting.

File: connector/src/main/java/org/geysermc/connector/utils/WebUtils.java [hunk 23]
```diff
@@ -132,4 +136,41 @@ private static String connectionToString(HttpURLConnection con) throws IOExcepti
...
+        http.setRequestProperty("Content-Type", "application/x-www-form-urlencoded; charset=UTF-8");
+        http.setRequestProperty("User-Agent", userAgent + "/" + version + "/" + mcversion);
+        http.connect();
+        try (OutputStream os = http.getOutputStream()) {
+            os.write(out);
+        }
+        String br = new BufferedReader(new InputStreamReader(http.getInputStream()))
+                .lines()
+                .collect(Collectors.joining());
+        ObjectMapper om = new ObjectMapper();
```
Comment (Camotoy) [inline connector/src/main/java/org/geysermc/connector/utils/WebUtils.java:171]: You should be able to re-use the `GeyserConnector.JSON_MAPPER` variable.

**Review by Konicai — CHANGES_REQUESTED** [review]

File: connector/src/main/java/org/geysermc/connector/utils/WebUtils.java (hunk 23 above)
Comment (Konicai) [reply connector/src/main/java/org/geysermc/connector/utils/WebUtils.java:171]: Also I would rename `String br` to something else, since its the json as a String, the variable isnt a buffered reader

File: connector/src/main/java/org/geysermc/connector/dump/DumpInfo.java [hunk 24]
```diff
@@ -188,6 +194,18 @@ public DumpInfo() {
...
 
+    @Getter
+    public static class LogsInfo {
+        private String link;
+
+        public LogsInfo() {
+            try {
+                this.link = WebUtils.postLogs(GeyserConnector.getInstance().getBootstrap().getLogsPath());
+            } catch (IOException ignored) {
+            }
```
Comment (rtm516) [inline connector/src/main/java/org/geysermc/connector/dump/DumpInfo.java:205]:
```suggestion
            } catch (IOException ignored) { }
```

**Review by rtm516 — CHANGES_REQUESTED** [review]

File: connector/src/main/java/org/geysermc/connector/utils/WebUtils.java [hunk 25]
```diff
@@ -132,4 +136,38 @@ private static String connectionToString(HttpURLConnection con) throws IOExcepti
...
+    public static String postLogs(Path log) throws IOException {
+        // Connect to api
+        URL url = new URL("https://api.mclo.gs/1/log");
+        URLConnection con = url.openConnection();
+        HttpURLConnection http = (HttpURLConnection) con;
+        http.setRequestMethod("POST");
+        http.setDoOutput(true);
+        // Convert log to application/x-www-form-urlencoded
+        String content = "content=" + new BufferedReader(new InputStreamReader(Files.newInputStream(log.toRealPath()))).lines().collect(Collectors.joining("\n"));
+        byte[] out = content.getBytes(StandardCharsets.UTF_8);
```
[the remark this answers is not in the document]
Comment (Konicai) [reply connector/src/main/java/org/geysermc/connector/utils/WebUtils.java:154]: also this
https://github.com/aternosorg/mclogs-java/blob/3b50129618f4599bfe653ac72ff06dd3c877f38f/src/main/java/gs/mclo/java/MclogsAPI.java#L35

**Review by Camotoy — APPROVED** [review]

**Comment by u9g** [comment]
Isn’t it like a big privacy invasion to upload a users private logs to a public host probably without their knowledge?

**Comment by Camotoy** [comment]
Logs will not be uploaded unless they are specified as a command option in our dump command.

**Comment by u9g** [comment]
> Logs will not be uploaded unless they are specified as a command option in our dump command.

So some new user comes to the discord for help and someone tells them run this command, and boom, all your data is uploaded to a public host?

**Comment by Camotoy** [comment]
The site we upload logs to censors out IPs which, in my opinion, is the only personally identifiable data in a log. Everything else is needed and would be asked for anyway already. If you have a concern about the way this will be integrated into our support processes, I recommend you head on over to https://github.com/GeyserMC/GeyserDiscordBot/pull/161 which is where the information will be presented from our bot.

**Comment by Konicai** [comment]
Also, if anyone wants to review the information before uploading to dump.geysermc.org, they can include `offline` in the dump command arguments to create the information in a local file instead.

Data uploaded to dump.geysermc.org is publicly accessible, but dump IDs/links are only shared with the user running the command

**Comment by u9g** [comment]
> Also, if anyone wants to review the information before uploading to dump.geysermc.org, they can include `offline` in the dump command arguments to create the information in a local file instead.
> 
> 
> 
> Data uploaded to dump.geysermc.org is publicly accessible, but dump IDs/links are only shared with the user running the command

This PR isnt about uploading to dump.geysermc.org is it? its about uploading to mclo.gs, which is not under geyser control at all

**Comment by u9g** [comment]
> The site we upload logs to censors out IPs which, in my opinion, is the only personally identifiable data in a log. Everything else is needed and would be asked for anyway already without this PR. If you have a concern about the way this will be integrated into our support processes, I recommend you head on over to https://github.com/GeyserMC/GeyserDiscordBot/pull/161 which is where the information will be presented from our bot.

You can't account for plugins that leak other sensitive information, and in my opinion any blind uploading of logs like this is very bad, it just reeks of data leaks waiting to happen on a site that gives you no contol because its not under geysermc control.

**Comment by u9g** [comment]
RE the discord bot, that's not my main concern, my main concern is blind uploading of logs data which may include sensitive information to a host that to my knowledge doesn't allow deleting the logs

**Comment by Camotoy** [comment]
We can always change to our own log sharing system if needed, or you can suggest something to https://github.com/aternosorg/mclogs. The problem is, of course, that an uploading system with no account means you can't easily delete the data.

> You can't account for plugins that leak other sensitive information

As far as I know, any Java Edition-related support server will ask for complete logs, so this isn't a problem that is solely on us.

> my main concern is blind uploading of logs data which may include sensitive information

That is why I suggested looking at the PR - the `logsdump.tag` contains the file that will be shown to people when someone runs the `!!logsdump` command in our server to request logs.

You still mention sensitive data, but I've never seen personal data in logs ever being an issue, other than IP addresses (and like one complaint about UUIDs, but if we know your username we can find your UUID).

**Comment by lukeeey** [comment]
> RE the discord bot, that's not my main concern, my main concern is blind uploading of logs data which may include sensitive information to a host that to my knowledge doesn't allow deleting the logs

The logs are uploaded **only** if the user *specifies* it, so i don’t see the problem.

You have to tell it to upload the logs. Also, I believe Aternos to be a trustworthy company. I see no reason not to trust them, as i see no reason not to trust pastebin. There is no need for us to reinvent the wheel with everything when Aternos already provide a great service.

**Comment by lukeeey** [comment]
Also logs automatically delete after a certain period of time i believe, which is something you won’t get on most paste sites.
